### PR TITLE
Native filtering/sorting v2

### PR DESCRIPTION
**Pull request by plyhun**
## Abstract

Filter / sort requests natively by database, speeding up the performance.

### General

* [x] Added abstract that describes the change
* [x] Added changelog entry to `/CHANGELOG.adoc`
* [x] Ensured that the change is covered by tests
* [x] Ensured that the change is documented in the docs

### On API Changes

* [x] Checked if the changes are breaking or not
* [x] Added GraphQL API if applicable
* [x] Added Elasticsearch mapping if applicable

### CODE REVIEW

File: bom/pom.xml [hunk 1]
```diff
@@ -692,7 +692,7 @@
 			<dependency>
 				<groupId>com.gentics.graphqlfilter</groupId>
 				<artifactId>graphql-java-filter</artifactId>
-				<version>2.0.0</version>
+				<version>2.0.1-SNAPSHOT</version>
```
Comment (plyhun) [inline bom/pom.xml:695]: TODO fix to release when ready.
https://github.com/gentics/graphql-java-filter/pull/5

File: common/src/main/java/com/gentics/mesh/core/link/WebRootLinkReplacerImpl.java [hunk 2]
```diff
@@ -175,7 +176,7 @@ public String resolve(InternalActionContext ac, String branch, ContainerType edg
 					//if there is a S3 field and we can do the link resolving with S3 from the configuration then we should return the presigned URL
 					if (Objects.isNull(linkResolver) || linkResolver.equals("s3")) {
 						String fieldName = s3binaryFieldSchema.getName();
-						return Optional.ofNullable(nullableGraphFieldContainer.getS3Binary(fieldName).getBinary());
+						return Optional.ofNullable(nullableGraphFieldContainer.getS3Binary(fieldName)).map(S3HibBinaryField::getBinary);
```
Comment (plyhun) [inline common/src/main/java/com/gentics/mesh/core/link/WebRootLinkReplacerImpl.java:179]: NPE fix

File: mdm/api/src/main/java/com/gentics/mesh/parameter/impl/PagingParametersImpl.java [hunk 3]
```diff
@@ -67,7 +69,7 @@ public PagingParametersImpl(int page) {
 	 *            Per page count
 	 */
 	public PagingParametersImpl(int page, Long perPage) {
-		this(page, perPage, "uuid", SortOrder.ASCENDING);
+		this(page, perPage, StringUtils.EMPTY, SortOrder.UNSORTED);
```
Comment (plyhun) [inline mdm/api/src/main/java/com/gentics/mesh/parameter/impl/PagingParametersImpl.java:72]: The default sorting should be turned off.

File: rest-model/src/main/java/com/gentics/mesh/parameter/FilteringParameters.java [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.gentics.mesh.parameter;
+
+public interface FilteringParameters {
```
Comment (plyhun) [inline rest-model/src/main/java/com/gentics/mesh/parameter/FilteringParameters.java:3]: Placeholder WIP.

File: rest-model/src/main/java/com/gentics/mesh/parameter/PagingParameters.java [hunk 5]
```diff
@@ -58,56 +54,4 @@ default PagingParameters setPerPage(Long perPage) {
 		}
 		return this;
 	}
-
```
Comment (plyhun) [inline rest-model/src/main/java/com/gentics/mesh/parameter/PagingParameters.java:61]: Sorting parameters cut apart for distinct usage in the core.

File: verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/AbstractTypeProvider.java [hunk 6]
```diff
@@ -573,6 +593,26 @@ protected DynamicStreamPageImpl<NodeContent> applyNodeFilter(DataFetchingEnviron
 		PagingParameters pagingInfo = getPagingInfo(env);
 		GraphQLContext gc = env.getContext();
 
+		NativeQueryFiltering nativeQueryFiltering = options.getGraphQLOptions().getNativeQueryFiltering();
```
Comment (plyhun) [inline verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/AbstractTypeProvider.java:596]: Heavy WIP here.

File: verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/AbstractTypeProvider.java [hunk 7]
```diff
@@ -563,20 +636,121 @@ protected DynamicStreamPageImpl<NodeContent> fetchFilteredNodes(DataFetchingEnvi
...
+				}
+				javaFilter = filterProvider.createPredicate(filterArgument);
+				break;
+			}
+		}
+		return Pair.of(javaFilter, maybeNativeFilter
+				.or(() -> Optional.of(PersistingRootDao.shouldPage(getPagingInfo(env)))
+						// We force native filtering, if paging is asked
+						.filter(b -> b && (NativeFilter.ONLY.equals(envNativeFilter) || NativeQueryFiltering.ALWAYS.equals(nativeQueryFiltering)))
+						.map(yes -> Comparison.dummy(true))));
```
Comment (npomaroli) [inline verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/AbstractTypeProvider.java:697]: This does not compile (with the currently checked in version of graphql-filter)

**Review by npomaroli — CHANGES_REQUESTED** [review]

File: verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/AbstractTypeProvider.java (hunk 7 above)
Comment (plyhun) [reply verticles/graphql/src/main/java/com/gentics/mesh/graphql/type/AbstractTypeProvider.java:697]: Argh. This is a '1 = 1' and '1 <> 1' operator.

**Review by npomaroli — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/gentics/mesh/etc/config/NativeQueryFiltering.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	private NativeQueryFiltering(int level) {
+		this.level = level;
+	}
+
+	/**
+	 * Get the filtering int value.
+	 * 
+	 * @return
+	 */
+	public int getLevel() {
```
[the remark this answers is not in the document]
Comment (plyhun) [reply api/src/main/java/com/gentics/mesh/etc/config/NativeQueryFiltering.java:37]: It's the same with `ResampleFilter` option enum. Either a level or a name can be used.

File: rest-client/src/main/java/com/gentics/mesh/parameter/client/PagingParametersImpl.java [hunk 9]
```diff
@@ -5,6 +5,6 @@
 /**
  * @see PagingParameters
  */
-public class PagingParametersImpl extends AbstractParameters implements PagingParameters {
+public class PagingParametersImpl extends SortingParametersImpl implements PagingParameters {
```
[the remark this answers is not in the document]
Comment (plyhun) [reply rest-client/src/main/java/com/gentics/mesh/parameter/client/PagingParametersImpl.java:8]: However the sorting has to work, so I'll check it.

File: bom/pom.xml (hunk 1 above)
Comment (plyhun) [reply bom/pom.xml:695]: Should it be `2.1.0` / `3.0.0` ?

File: mdm/api/pom.xml [hunk 10]
```diff
@@ -29,6 +24,9 @@
 			<groupId>com.hazelcast</groupId>
 			<artifactId>hazelcast</artifactId>
 		</dependency>
+		<dependency>
```
Comment (plyhun) [inline mdm/api/pom.xml:27]: Need to add a GraphQl type for `loadAll` below.

**Pull request merged**